src/libcharon/plugins/kernel_super/system_events/sl_system_events/sl_system_events.c

```diff
@@ -0,0 +1,165 @@
+#include "sl_system_events.h"
+
+#include "../libsystem_events/libsystem_events.h"
+
+#include <dlfcn.h>
+#include <stdlib.h>
+#include <string.h>
+#include <utils/debug.h>
+
+#define LIBSYSTEM_EVENTS_SO_PATH "/usr/local/lib/libsystem_events.so"
+
+typedef struct private_sl_system_events_t private_sl_system_events_t;
+
+struct private_sl_system_events_t {
+	sl_system_events_t public;
+	void *dl_handle;
+	struct {
+		typeof(libsystem_events_send_ipsec_tunnel_established)
+			*libsystem_events_send_ipsec_tunnel_established;
+		typeof(libsystem_events_send_ipsec_tunnel_establishment_failed)
+			*libsystem_events_send_ipsec_tunnel_establishment_failed;
+		typeof(libsystem_events_send_ipsec_tunnel_activated)
+			*libsystem_events_send_ipsec_tunnel_activated;
+		typeof(libsystem_events_send_ipsec_tunnel_deactivated)
+			*libsystem_events_send_ipsec_tunnel_deactivated;
+		typeof(libsystem_events_send_ipsec_tunnel_down) *libsystem_events_send_ipsec_tunnel_down;
+		typeof(
+			libsystem_events_send_ipsec_terminator_up) *libsystem_events_send_ipsec_terminator_up;
+		typeof(libsystem_events_send_ipsec_terminator_down)
+			*libsystem_events_send_ipsec_terminator_down;
+	} ops;
+};
+
+METHOD(system_events_t, sl_send_ipsec_tunnel_established, bool, private_sl_system_events_t *this,
```

Comment:
These methods should be named the same as the name in the interface without the struct name as a prefix (similarly to how in C++ you don't prefix class member functions names with the class's name. The `METHOD()` macro behaves the same).
So for example this method should be named `send_ipsec_tunnel_established` without the `sl_` prefix

Reply:
What about `destroy` function?
`METHOD(system_events_t, destroy, void, private_sl_system_events_t *this)` ?